include/zephyr/modem/pipelink.h

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) 2024 Trackunit Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/modem/pipe.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/sys/util.h>
+
+#ifndef ZEPHYR_MODEM_PIPELINK_
+#define ZEPHYR_MODEM_PIPELINK_
+
+/** Pipelink event */
+enum modem_pipelink_event {
+	/** Modem pipe has been connected and can be opened */
+	MODEM_PIPELINK_EVENT_CONNECTED = 0,
+	/** Modem pipe has been disconnected and can't be opened */
+	MODEM_PIPELINK_EVENT_DISCONNECTED,
+};
+
+/** @cond INTERNAL_HIDDEN */
+
+/** Forward declaration */
+struct modem_pipelink;
+
+/** @endcond */
+
+/**
+ * @brief Pipelink callback definition
+ * @param link Modem pipelink instance
+ * @param event Modem pipelink event
+ * @param user_data User data passed to modem_pipelink_attach()
+ */
+typedef void (*modem_pipelink_callback)(struct modem_pipelink *link,
+					enum modem_pipelink_event event,
+					void *user_data);
+
+/** @cond INTERNAL_HIDDEN */
+
+/** Pipelink structure */
+struct modem_pipelink {
+	struct modem_pipe *pipe;
+	modem_pipelink_callback callback;
+	void *user_data;
+	bool connected;
+	struct k_spinlock spinlock;
+};
+
+/** @endcond */
+
+/**
+ * @brief Attach pipelink to callback
+ * @param link Pipelink instance
+ * @param callback Pipelink callback
+ * @param user_data User data passed to pipelink callback
+ */
+void modem_pipelink_attach(struct modem_pipelink *link,
+			   modem_pipelink_callback callback,
+			   void *user_data);
+
+/**
+ * @brief Test if pipelink pipe is connected
+ * @param link Pipelink instance
+ * @retval true if pipe is connected
+ * @retval false if pipe is not connected
+ */
+bool modem_pipelink_is_connected(struct modem_pipelink *link);
+
+/**
+ * @brief Get pipe from pipelink
+ * @param link Pipelink instance
+ * @retval Pointer to pipe if pipelink has been initialized
+ * @retval NULL if pipelink has not been initialized
+ * @note Pipelink is initialized on first connected event
```

Comment:
What do you mean with this?

Reply:
Updated the note to `Pipelink is guaranteed initialized after first connected event`. The point being that if you wait for the first connected event, `modem_pipelink_get_pipe()` will not return NULL, so you don't have to check the return every time :)

Reply:
Okay. Though the first connected event doesn't do have anything to do with `link->pipe`, as it's set already and only in `modem_pipelink_init()`?

Reply:
Correct :) Maybe the note is implied? Just thought it may be helpful

Reply:
I rather find the note confusing in that case, talking about connected events while in practice they have nothing to do?

Reply:
I have removed the note :)